File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
       int n= knowledge.size();
       unordered_map<string,string>mp;
       for(int i=0;i<n;i++){
             mp[knowledge[i][0]]=knowledge[i][1];
       }   
       bool check=false;
       int np=s.size();
       string ans="";
        string sp="";
       for(int i=0;i<np;i++){
          
        if(s[i]!='(' && s[i]!=')' && check==false){
             ans+=s[i];
        }else if(s[i]=='('){
            sp="";
            check=true;
        }else if(s[i]==')'){
           
            if(mp.find(sp)!=mp.end()){
              
                ans+=mp[sp];
            }else{
                ans+='?';
            }
            sp="";
            check=false;
        }else if(s[i]!='(' && s[i]!=')' && check==true){
            sp+=s[i];
        }
       }
       return ans;
    }
};

File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
         int n= knowledge.size();
      HashMap<String,String>mp=new HashMap<>();
       for(int i=0;i<n;i++){
             mp.put(knowledge.get(i).get(0),knowledge.get(i).get(1));
       }   
       boolean check=false;
       int np=s.length();
       String ans="";
        String sp="";
       for(int i=0;i<np;i++){
          
        if(s.charAt(i)!='(' && s.charAt(i)!=')' && check==false){
             ans+=s.charAt(i);
        }else if(s.charAt(i)=='('){
            sp="";
            check=true;
        }else if(s.charAt(i)==')'){
           
            if(mp.containsKey(sp)){
              
                ans+=mp.get(sp);
            }else{
                ans+='?';
            }
            sp="";
            check=false;
        }else if(s.charAt(i)!='(' && s.charAt(i)!=')' && check==true){
            sp+=s.charAt(i);
        }
       }
       return ans;
    }
}